File: languages/family_trees/display_tree.js
// compiled from Hot Cocoa Lisp

var __, csv, format = function(f,v){var i=0;return f.replace(/~([a-zA-Z0-9_]*)~/g,function(_,k){if(k===''){k=i;i++;}if(v[k]===undefined){return'';}return v[k];})}, people, prompt, current__person;

// ;;; Sam Auciello | 2013 | languages Exam
// ;;; This program reads in people from people.csv constructs a linked family tree
// ;;; structure from it, then provides a console interface to navigate the tree.
// (def _ (require "underscore"))

__ = require("underscore");

// (def csv (require (compile "./csv.hcl")))

csv = require("./csv.js");

// (def people
// 	 (_.map (csv "people.csv")
// 			(# (person)
// 			   (when (= "-1" person.died) (set person.died "present"))
// 			   (_.extend person {
// 						 
// 				  getFather
// 				  (# () (get people (integer this.father)))
// 				  
// 				  getMother
// 				  (# () (get people (integer this.mother)))
// 				  
// 				  children
// 				  (# ()
// 					 (let (that this)
// 					   (_.filter people
// 								 (# (person)
// 									(contains? [ person.mother person.father ]
// 											   that.id)))))
// 				  
// 				  child
// 				  (# (n) (get (this.children) n))
// 				  
// 				  spouce
// 				  (# ()
// 					 (if (empty? (this.children)) undefined
// 					   (if (= this.id (get (this.children) 0).mother)
// 						   ((get (this.children) 0).getFather)
// 						 ((get (this.children) 0).getMother))))
// 				  
// 				  print
// 				  (# () (console.log (format "~~ ~~ (~~ - ~~)"
// 											 [ (get this "first_name")
// 											   (get this "last_name")
// 											   this.born this.died ] )))
// 				  
// 				  info
// 				  (# ()
// 					 (process.stdout.write "\n")
// 					 (this.print)
// 					 (process.stdout.write "\nMother: ")
// 					 (if (this.getMother) ((this.getMother).print)
// 					   (console.log "N/A"))
// 					 (process.stdout.write "Father: ")
// 					 (if (this.getFather) ((this.getFather).print)
// 					   (console.log "N/A"))
// 					 (when (not (empty? (this.children)))
// 					   (process.stdout.write "Spouce: ")
// 					   ((this.spouce).print)
// 					   (process.stdout.write "Children:\n")
// 					   (for (child (this.children)) (child.print)))) } ))))

people = __.map(csv("people.csv"), (function(person) { ((("-1" === person.died)) && (function() { (person.died = "present"); }).call(this)); return __.extend(person, { getFather: (function() {  return people[Math.floor(parseFloat(this.father))]; }), getMother: (function() {  return people[Math.floor(parseFloat(this.mother))]; }), children: (function() {  return (function(that) { return __.filter(people, (function(person) {  return ([person.mother, person.father].indexOf(that.id) !== -1); })); }).call(this, this); }), child: (function(n) {  return this.children()[n]; }), spouce: (function() {  return ((this.children() === null || (this.children()).length === 0) ? undefined : (((this.id === this.children()[0].mother)) ? this.children()[0].getFather() : this.children()[0].getMother())); }), print: (function() {  return console.log(format("~~ ~~ (~~ - ~~)", [this["first_name"], this["last_name"], this.born, this.died])); }), info: (function() {var _i0_, child; process.stdout.write("\n"); this.print(); process.stdout.write("\nMother: "); (this.getMother() ? this.getMother().print() : console.log("N/A")); process.stdout.write("Father: "); (this.getFather() ? this.getFather().print() : console.log("N/A")); return ((! (this.children() === null || (this.children()).length === 0)) && (function() { process.stdout.write("Spouce: "); this.spouce().print(); process.stdout.write("Children:\n"); for (_i0_ = 0; _i0_ < this.children().length; _i0_++) { child = this.children()[_i0_]; child.print(); }; }).call(this)); }) }); }));

// (def prompt "\nEnter a relative to navigate to (mother, father, spouce, child_n): ")

prompt = "\nEnter a relative to navigate to (mother, father, spouce, child_n): ";

// (var current_person (get people 0))

current__person = people[0];

// (current_person.info)

current__person.info();

// (process.stdout.write prompt)

process.stdout.write(prompt);

// (process.stdin.resume)

process.stdin.resume();

// (process.stdin.setEncoding "utf8")

process.stdin.setEncoding("utf8");

// (process.stdin.on "data"
// 				  (# (input)
// 					 (let (command (input.trim))
// 					   (cond
// 						((= command "quit") (process.exit))
// 						((= command "mother")
// 						 (if (nil? (current_person.getMother))
// 							 (console.log "\n! no mother available")
// 						   (set current_person (current_person.getMother))))
// 						((= command "father")
// 						 (if (nil? (current_person.getFather))
// 							 (console.log "\n! no father available")
// 						   (set current_person (current_person.getFather))))
// 						((= command "spouce")
// 						 (if (nil? (current_person.spouce))
// 							 (console.log "\n! no spouce available")
// 						   (set current_person (current_person.spouce))))
// 						(((re "^child_\d+$").exec command)
// 						 (let (child (current_person.child
// 									  (command.replace "child_" "")))
// 						   (if (nil? child)
// 							   (console.log (cat "\n! no child "
// 												 (command.replace "child_" "")
// 												 "  available"))
// 							 (set current_person child))))
// 						(true (console.log (cat "\n! invalid command: "
// 												command )))))
// 					 (current_person.info)
// 					 (process.stdout.write prompt)))

process.stdin.on("data", (function(input) { (function(command) { return (((command === "quit")) ? process.exit() : ((command === "mother")) ? ((current__person.getMother() === null || current__person.getMother() === undefined) ? console.log("\n! no mother available") : (current__person = current__person.getMother())) : ((command === "father")) ? ((current__person.getFather() === null || current__person.getFather() === undefined) ? console.log("\n! no father available") : (current__person = current__person.getFather())) : ((command === "spouce")) ? ((current__person.spouce() === null || current__person.spouce() === undefined) ? console.log("\n! no spouce available") : (current__person = current__person.spouce())) : (new RegExp("^child_\\d+$", "")).exec(command) ? (function(child) { return ((child === null || child === undefined) ? console.log(("\n! no child " + command.replace("child_", "") + "  available")) : (current__person = child)); }).call(this, current__person.child(command.replace("child_", ""))) : true ? console.log(("\n! invalid command: " + command)) : undefined); }).call(this, input.trim()); current__person.info(); return process.stdout.write(prompt); }));

File: languages/family_trees/display_tree.c
/* Sam Auciello | 2013 | languages Exam
 * This program reads in people from people.csv constructs a linked family tree
 * structure from it, then provides a console interface to navigate the tree.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct _person *person;
typedef struct _people *people;

struct _person {
  char* first_name;
  char* last_name;
  int father_id; // the index of the parent in the csv file
  int mother_id;
  person father; // a link the the actual parent struct
  person mother;
  people children; // a list of children
  int born;
  int died;
};

// lists of people are stored as a linked list
struct _people {
  person car; // the first element of the list
  people cdr; // the sublist following the first element
};

// allocate space for a new person
person new_person() {
  person result = malloc(sizeof(struct _person));
  result->first_name = malloc(sizeof(char) * 40);
  result->last_name = malloc(sizeof(char) * 40);
  return result;
}

// add a person to people linked list.  This should be used with something like:
// people list;
// person jane;
// list = people_push(list, jane);
people people_push(people list, person p) {
  people result = malloc(sizeof(struct _people));
  result->car = p;
  result->cdr = list;
  return result;
}

// print out the person in form "Jane Doe (1902 - 1987)\n"
// prints "N/A\n" for the NULL person
void print_person(person p) {
  if (p == NULL) {
	printf("N/A\n");
  } else {
	printf("%s %s (%i - ", p->first_name, p->last_name, p->born);
	if (p->died < 0) {
	  printf("present)\n");
	} else {
	  printf("%i)\n", p->died);
	}
  }
}

// print each person in the list
void print_people(people p) {
  for (; p != NULL; p = p->cdr) {
	print_person(p->car);
  }
}

// find the spouce of the person if there is one
person spouce(person p) {
  if (p->children == NULL) {
	return NULL;
  }
  if (p->children->car->father == p) {
	return p->children->car->mother;
  }
  return p->children->car->father;
}

// print the person, their parents, their spouce, and their children
void person_info(person p) {
  printf("\n");
  print_person(p);
  printf("\nMother: ");
  print_person(p->mother);
  printf("Father: ");
  print_person(p->father);
  if (p->children != NULL) {
	printf("Spouce: ");
	print_person(spouce(p));
	printf("Children:\n");
	print_people(p->children);
  }
}

// find the person corresponding to the given index into the people list
person nth_person(people list, int n) {
  if (list == NULL) {
	return NULL;
  }
  people p = list;
  int i = 0;
  while (p->cdr != NULL && i < n) {
	i++;
	p = p->cdr;
  }
  if (i == n) {
	return p->car;
  }
  return NULL;
}

// generate a people list from a .csv file
people load_people(FILE* csv_file) {
  people result = NULL, last_person, p;
  char* line_buf = malloc(sizeof(char) * 80);
  fgets(line_buf, 80, csv_file); // ignore header
  while (fgets(line_buf, 80, csv_file)) {
	
	// each line
	
	// make sure last_person points to an empty people list at the end of the
	// linked result list
	if (result == NULL) {
	  result = last_person = malloc(sizeof(struct _people));
	} else {
	  last_person->cdr = malloc(sizeof(struct _people));
	  last_person = last_person->cdr;
	}
	
	// add the next person at the end
	last_person->car = new_person();
	sscanf(line_buf, "%*i,%[^,],%[^,],%i,%i,%i,%i",
		  last_person->car->first_name,
		  last_person->car->last_name,
		  &last_person->car->father_id,
		  &last_person->car->mother_id,
		  &last_person->car->born,
		  &last_person->car->died);
  }
  
  // create links to parents and children
  for (p = result; p != NULL; p = p->cdr) {
	if (p->car->father_id >= 0) {
	  p->car->father = nth_person(result, p->car->father_id);
	  p->car->father->children = people_push(p->car->father->children, p->car);
	}
	if (p->car->mother_id >= 0) {
	  p->car->mother = nth_person(result, p->car->mother_id);
	  p->car->mother->children = people_push(p->car->mother->children, p->car);
	}
  }
  return result;
}

int main() {
  
  // setup
  FILE* f = fopen("people.csv", "r");
  people list = load_people(f);
  person current_person = list->car;
  char* command = malloc(sizeof(char) * 80);
  
  // command prompt loop
  while (true) {
	person_info(current_person);
	printf("\nEnter a relative to navigate to (mother, father, spouce, child_n): ");
	gets(command);
	if (strcmp(command, "quit") == 0) {
	  break;
	} else if (strcmp(command, "mother") == 0) {
	  if (current_person->mother == NULL) {
		printf("\n! no mother available\n");
	  } else {
		current_person = current_person->mother;
	  }
	} else if (strcmp(command, "father") == 0) {
	  if (current_person->father == NULL) {
		printf("\n! no father available\n");
	  } else {
		current_person = current_person->father;
	  }
	} else if (strcmp(command, "spouce") == 0) {
	  if (spouce(current_person) == NULL) {
		printf("\n! no spouce available\n");
	  } else {
		current_person = spouce(current_person);
	  }
	} else if (strncmp(command, "child_", 6) == 0) {
	  int n = atoi(&command[6]);
	  if (nth_person(current_person->children, n) == NULL) {
		printf("\n! no child %i available\n", n);
	  } else {
		current_person = nth_person(current_person->children, n);
	  }
	} else {
	  printf("\n! invalid command: %s\n", command);
	}
  }
  
  return 0;
}
